Fix 'meson introspect --target-files' for a custom target

jon-turney opened the pull request:
Add testcase for, and fix `meson introspect --target-files` for a custom target, broken since 0.26.0 (PR #225)
For a CustomTarget, sources is a list of filenames as strings, not File objects.

Closes #2783

jon-turney commented on run_unittests.py:
```diff
@@ -1737,6 +1739,38 @@ def test_templates(self):
+            'prog@exe': ['prog.c'],
+        }
+        t_intro = self.introspect('--targets')
+        self.assertEqual([t['id'] for t in t_intro], list(expected.keys()))
```
Yes.  In fact, since assertCountEqual() is needed here, it seems fine to use the iterable directly

jon-turney commented:
> The "correct" solution would be to convert custom target sources to Files with self.source_strings_to_files. Can you test if this patch works for you:

Yes, that also works for custom targets in mesa.

I'm not sure about breaking `vala/9 gir` like that, though. Generating a typeliib like that [is documented](http://mesonbuild.com/Vala.html#gobject-introspection), so that's just trading one regression for another...

jpakkane commented:
This is my bad, I have been totally asleep at the wheel to accept that. Manually building paths to files is _bad_. We should never promote that in our tests or official docs. I really want to fix this properly, though, because this kind of string inconsistency is a big maintenance burden.

jon-turney commented:
> This is my bad, I have been totally asleep at the wheel to accept that. Manually building paths to files is bad. We should never promote that in our tests or official docs. I really want to fix this properly, though, because this kind of string inconsistency is a big maintenance burden.

Ok.  This PR just contains the test case now.

nirbheek commented:
> This is my bad, I have been totally asleep at the wheel to accept that. Manually building paths to files is bad.

Funnily enough, somehow I missed that too, so I share the blame.

jon-turney commented:
Rebased to fix conflicts.

I added something to handle non-file inputs in a backwards compatble way, with a warning:

```console
jon@tambora /meson/test cases/vala/9 gir
$ rm -r _build ; /meson/meson.py _build
[...]
WARNING: Custom target input '/wip/meson/test cases/vala/9 gir/_build/Foo-1.0.gir' can't be converted to a File object.
This will become a hard error in the future.
```

This warning should probably say something more about what's probably wrong and how to remedy things.

>  Generating a typelib like that [is documented](http://mesonbuild.com/Vala.html#gobject-introspection)

I guess a defect needs to be created about how vala libraries don't expose the `.gir` as an output, or something like that.

jpakkane commented:
I created the bug as #4481.
